### PR TITLE
Add documentation for symbols and for..of statements

### LINKED ISSUE #33: iterators/generators
Iterables in ES6, Symbol.iterator, for..of, using iterators, generators

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #33

First stab at iterators and generators. I still need to add documentation for generators.

### CODE REVIEW

File: pages/Iterators and Generators.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+Symbols are immutable, and unique.
+
+```ts
+var sym2 = Symbol("key"); 
+var sym3 = Symbol("key"); 
+
+sym2 === sym3; // false, symbols are unique
+```
+
+Just like strings, symbols can be used as keys for object properties.
```
Comment (yuit) [inline pages/Iterators and Generators.md]: Similar to

File: pages/Iterators and Generators.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+```ts
+let sym = Symbol();
+
+let obj = {};
+
+obj[sym] = "value";
+console.log(obj[sym]); // "value" 
+```
+
+Combined with computed property declarations, symbols can be used to declare object properties and class members.
```
Comment (yuit) [inline pages/Iterators and Generators.md]: Symbols can also be combined with computed property to declare object properties and class members.

File: pages/Iterators and Generators.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+In addition to user-defined symbols, there are well-known built-in symbols.
+Built-in symbols are used to represent internal language behaviors.
+
+Here is a list of well-known symbols:
+
+* Symbol.hasInstance  
+  A method that determines if a constructor object recognizes an object as one of the constructor’s instances. Called by the semantics of the instanceof operator. 
+
+* isConcatSpreadable
```
Comment (yuit) [inline pages/Iterators and Generators.md]: Symbol.isConcatSpreadable

File: pages/Iterators and Generators.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+In addition to user-defined symbols, there are well-known built-in symbols.
+Built-in symbols are used to represent internal language behaviors.
+
+Here is a list of well-known symbols:
+
+* Symbol.hasInstance  
+  A method that determines if a constructor object recognizes an object as one of the constructor’s instances. Called by the semantics of the instanceof operator. 
+
+* isConcatSpreadable  
+  A Boolean value that if true indicates that an object should flatten to its array elements by Array.prototype.concat.
```
Comment (yuit) [inline pages/Iterators and Generators.md]: A Boolean value indicating that an object should be ....

File: pages/Iterators and Generators.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+* Symbol.toPrimitive  
+  A method that converts an object to a corresponding primitive value.
+  Called by the `ToPrimitive` abstract operation.
+
+* Symbol.toStringTag  
+  A String value that is used in the creation of the default string description of an object.
+  Called by the built-in method `Object.prototype.toString`.
+
+* Symbol.unscopables  
+  An Object whose own property names are property names that are excluded from the 'with' environment bindings of the associated objects.
```
Comment (yuit) [inline pages/Iterators and Generators.md]: An Object whose property names

File: pages/Iterators and Generators.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+```ts
+let someArray = [1, "string", false];
+
+for (let entry of someArray) {
+    console.log(item); // 1, "string", false
+}
+``` 
+
+### `for..of` vs. `for..in` statements
```
Comment (yuit) [inline pages/Iterators and Generators.md]: should ## `for..of` vs. `for..in` statements

File: pages/Iterators and Generators.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    console.log(pet); // "Cat", "Dog", "Hamster"
+}
+```
+
+### Code generation
+
+#### Targeting ES5 and ES3
+
+When targeting an ES5 or ES3, iterators are only allowed on values of `Array` type. 
+It is an error to use `for..of` loops on non-Array values, even it they implement the `Symbol.iterator` property.
```
Comment (yuit) [inline pages/Iterators and Generators.md]: even though the non-Array values implement the `Symbol.iterator` property

**Comment by yuit** [comment]
should symbol have it own section and then we can link Symbol section from iterator for more info? I don't think I will know to look into iterator and generator to read about symbol

File: pages/Iterators and Generators.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  A String value that is used in the creation of the default string description of an object.
+  Called by the built-in method `Object.prototype.toString`.
+
+* Symbol.unscopables  
+  An Object whose own property names are property names that are excluded from the 'with' environment bindings of the associated objects.
+
+# Iterables
+
+An object is deemed iterable if it has an implementation for the `Symbol.iterator` property. 
+Some built-int types like `Array`, `Map`, `Set`, `String`, `Int32`, `Unint32`, etc. have their `Symbol.iterator` property already implemented.
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: typos: built-in and Uint32. Also I couldn't find documentation on Uint32, only Uint32Array. Does Uint32 exist in javascript?

File: pages/Iterators and Generators.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
+# Symbols
+
+Starting with ECMAScript 2015, `symbols` are a primitive data type, just like `number` and `string`.
```
Comment (DanielRosenwasser) [inline pages/Iterators and Generators.md]: `symbol` is

File: pages/Iterators and Generators.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+`symbol` values are created by calling the `Symbol` constructor.
+
+```ts
+var sym1 = Symbol(); 
+
+var sym2 = Symbol("key"); // optional string key
```
Comment (DanielRosenwasser) [inline pages/Iterators and Generators.md]: `let`

File: pages/Iterators and Generators.md (hunk 1 above)
Comment (DanielRosenwasser) [inline pages/Iterators and Generators.md]: Just like is fine, and it should be strings.

File: pages/Iterators and Generators.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+for (let i in list) {
+   console.log(i); // "0", "1", "2", 
+}
+
+for (let i of list) {
+   console.log(i); // "4", "5", "6"
+```
+
+Another distinction is that `for..in` operates on any object; it serves as a way to inspect properties on this object. 
+`for..in` on the other hand, is mainly interested in values of iterable objects. Built-in objects like `Map` and `Set` implement `Symbol.iterator` property allowing access to stored values.
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: Should be for-of, and missing a comma: "`for..of`, on the other hand, "

File: pages/Iterators and Generators.md (hunk 9 above)
Comment (DanielRosenwasser) [inline pages/Iterators and Generators.md]: Or just "symbols are" and have a second sentence that says "All symbols are named by the type `symbol`." or something like that

File: pages/Iterators and Generators.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+## Well-known Symbols
+
+In addition to user-defined symbols, there are well-known built-in symbols.
+Built-in symbols are used to represent internal language behaviors.
+
+Here is a list of well-known symbols:
+
+* Symbol.hasInstance
```
Comment (DanielRosenwasser) [inline pages/Iterators and Generators.md]: These should all be surrounded by backticks

File: pages/Iterators and Generators.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+## `for..of` statements
+
+`for..of` loops over an iterable object, invoking the `Symbol.iterator` property on the object.
+Here is a simple `for..of` loop on an array:
+
+```ts
+let someArray = [1, "string", false];
+
+for (let entry of someArray) {
+    console.log(item); // 1, "string", false
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: should be `console.log(entry)`

File: pages/Iterators and Generators.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+}
+``` 
+
+### `for..of` vs. `for..in` statements
+
+Both `for..of` and `for..in` statements iterate over lists; the values iterated on are different though, `for..in` returns a list of keys *keys* on the object being iterated, whereas `for..of` returns a list of *values* of the numeric properties of the object being iterated.
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: duplicate word "keys *keys*"

File: pages/Iterators and Generators.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+var numbers = [1, 2, 3];
+for (var _i = 0; _i < numbers.length; _i++) {
+    var num = numbers[_i];
+    console.log(num);
+}
+```
+
+#### Targeting ECMAScript 2015 and higher
+
+When targeting an ECMAScipt 2015-compilaient engine, the compiler will generate `for..of` loops to target the built-in iterator implementation in the engine.
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: typo:compliant

File: pages/Iterators and Generators.md (hunk 2 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 3 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 4 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 5 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: i think "own property" here serves a purpose.

File: pages/Iterators and Generators.md (hunk 6 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: i meant it as a sub topic of the `for..of` one

File: pages/Iterators and Generators.md (hunk 7 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 8 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: that was wrong. i meant `Int32Array` and `Uint32Array`. fixed.

File: pages/Iterators and Generators.md (hunk 9 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done

File: pages/Iterators and Generators.md (hunk 11 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: i do not know.. @DanielRosenwasser how do you want it to be? i have used `..` everywhere when referring to `for..of` and `for..in` loops.

File: pages/Iterators and Generators.md (hunk 12 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 13 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: fixed.

File: pages/Iterators and Generators.md (hunk 14 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: done.

File: pages/Iterators and Generators.md (hunk 15 above)
Comment (mhegazy) [inline pages/Iterators and Generators.md]: fixed

File: pages/Iterators and Generators.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
+# Iterables
+
+An object is deemed iterable if it has an implementation for the [`Symbol.iterator`](Symbols.md#symboliterator) property.
+Some built-in types like `Array`, `Map`, `Set`, `String`, `Int32Array`, `Unint32Array`, etc. have their `Symbol.iterator` property already implemented.
```
Comment (sandersn) [inline pages/Iterators and Generators.md]: UInt32Array is still a typo.

**Comment by sandersn** [comment]
Still one typo. Otherwise :+1:

**Pull request merged**